cs/VowpalWabbitAsync.cs

```diff
@@ -9,14 +18,29 @@
 
 namespace VW
 {
+    /// <summary>
+    /// An async wrapper VW supporting data ingest using declarative serializer infrastructure used with <see cref="VowpalWabbitThreadedLearning"/>.
+    /// </summary>
+    /// <typeparam name="TExample">The user type to be serialized.</typeparam>
     public class VowpalWabbitAsync<TExample> : IDisposable
     {
+        /// <summary>
+        /// The owning manager.
+        /// </summary>
         private VowpalWabbitThreadedLearning manager;
 
+        /// <summary>
+        /// The serializers are not thread-safe. Thus we need to allocate one for each VW instance.
+        /// </summary>
         private VowpalWabbitSerializer<TExample>[] serializers;
 
         internal VowpalWabbitAsync(VowpalWabbitThreadedLearning manager)
         {
+            Contract.Requires(manager != null);
```

Comment:
Requires are preconditions (e.g. manager != null)
Ensures are postconditions that must be met on method exit.

From: Luong Hoang [mailto:notifications@github.com]
Sent: Monday, September 14, 2015 2:59 PM
To: JohnLangford/vowpal_wabbit <vowpal_wabbit@noreply.github.com>
Cc: Markus Cozowicz <marcozo@microsoft.com>
Subject: Re: [vowpal_wabbit] Documentation (#754)


In cs/VowpalWabbitAsync.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fJohnLangford%2fvowpal_wabbit%2fpull%2f754%23discussion_r39432521&data=01%7c01%7cmarcozo%40064d.mgd.microsoft.com%7c78acfe0bf4494b2172c708d2bd367f98%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=utVyL2jDVncvNV5KPQFHSNyi7b6e3NonRU9a5PqRZ4c%3d>:

>          private VowpalWabbitSerializer<TExample>[] serializers;

>

>          internal VowpalWabbitAsync(VowpalWabbitThreadedLearning manager)

>          {

> +            Contract.Requires(manager != null);

what's the difference between Requires and Ensures?

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fJohnLangford%2fvowpal_wabbit%2fpull%2f754%2ffiles%23r39432521&data=01%7c01%7cmarcozo%40064d.mgd.microsoft.com%7c78acfe0bf4494b2172c708d2bd367f98%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=qWyr00I%2fazTXCJ9LDBF3S9R8U0xKf8DbuO8u9xU1vj4%3d>.